pkg/handlers/defaulthandler.go

```diff
@@ -0,0 +1,245 @@
+// it either decompresses or extracts the contents directly, sending data to the provided channel.
+// Returns an error if the archive cannot be processed due to issues like exceeding maximum depth or unsupported formats.
+func (h *DefaultHandler) openArchive(ctx logContext.Context, depth int, reader io.Reader, archiveChan chan []byte) error {
+	if common.IsDone(ctx) {
+		return ctx.Err()
+	}
+
+	if depth > maxDepth {
+		return ErrMaxDepthReached
+	}
+
+	format, arReader, err := archiver.Identify("", reader)
+	switch {
+	case err == nil:
+		// Continue with the rest of the code.
+	case errors.Is(err, archiver.ErrNoMatch):
+		if depth > 0 {
+			// If openArchive is called on an already extracted/decompressed file and the depth is greater than 0,
+			// it means we are at least 1 layer deep in the archive. In this case, we should handle the content
+			// as non-archive data by calling handleNonArchiveContent.
+			return h.handleNonArchiveContent(ctx, arReader, archiveChan)
+		}
+		// If openArchive is called on the root (depth == 0) and we can't identify the format,
+		// it means we can't handle the content at all. Return the archiver.ErrNoMatch error.
+		return err
+	default:
+		// Some other error occurred.
+		return fmt.Errorf("error identifying archive: %w", err)
+	}
+
+	switch archive := format.(type) {
+	case archiver.Decompressor:
+		// Decompress tha archive and feed the decompressed data back into the archive handler to extract any nested archives.
+		compReader, err := archive.OpenReader(arReader)
+		if err != nil {
+			return err
+		}
+
+		defer compReader.Close()
+
+		return h.openArchive(ctx, depth+1, compReader, archiveChan)
+	case archiver.Extractor:
+		return archive.Extract(logContext.WithValue(ctx, depthKey, depth+1), arReader, nil, h.extractorHandler(archiveChan))
+	default:
+		return fmt.Errorf("unknown archive type: %s", format.Name())
+	}
+}
+
+// extractorHandler creates a closure that handles individual files extracted by an archiver.
+// It logs the extraction, checks for cancellation, and decides whether to skip the file based on its name or type,
+// particularly for binary files if configured to skip. If the file is not skipped, it recursively calls openArchive
+// to handle nested archives or to continue processing based on the file's content and depth in the archive structure.
+func (h *DefaultHandler) extractorHandler(archiveChan chan []byte) func(context.Context, archiver.File) error {
+	return func(ctx context.Context, file archiver.File) error {
+		lCtx := logContext.WithValues(
+			logContext.AddLogger(ctx),
+			"filename", file.Name(),
+			"size", file.Size(),
+		)
+		lCtx.Logger().V(5).Info("Handling extracted file.")
+
+		if common.IsDone(ctx) {
+			return ctx.Err()
+		}
+
+		depth := 0
+		if ctxDepth, ok := ctx.Value(depthKey).(int); ok {
+			depth = ctxDepth
+		}
+		if int(file.Size()) > maxSize {
```

Comment:
I might have just overlooked where this is handled, but I don't see where the total bytes being extracted is being counted. This was in the original archive handler to prevent zip bombs.

Reply:
Ah, so I think in the next PR this check will actually get removed since we'll use our bufferedFileReader to read the contents. That would prevent us from running into any memory issues. This was just a temporary change since we no longer use `ReadToMax`.